std/hash/keccaf.asm

```diff
@@ -0,0 +1,95 @@
+use std::array;
+
+// TODO
+let<T1, T2> routine: T1, T1, (T1, T1 -> T1), (T1, T1 -> T1), (T1, T1 -> T1) -> T2 =
+    |a, b, xor, and, or| a + b;
+
+// function definition
+let or: int, int -> int = |a, b| a | b;
+let and: int, int -> int = |a, b| a & b;
+let xor: int, int -> int = |a, b| a ^ b;
+
+// ln 9 - 12
+// TODO: check that these are good for word size 32
+let RHO: int[24] = [
+    1,  3,  6,  10, 15, 21, 28, 36, 45, 55, 2,  14,
+    27, 41, 56, 8,  25, 43, 62, 18, 39, 61, 20, 44
+];
+
+// ln 14 - 16
+// TODO: check that these are good for word size 32
+let PI: int[24] = [
+    10, 7,  11, 17, 18, 3, 5,  16, 8,  21, 24, 4,
+    15, 23, 19, 13, 12, 2, 20, 14, 22, 9,  6,  1
+];
+
+// ln 19 - 28
+// TODO: check that these are good for word size 32
+let RC: int[24] = [
+    0x0000000000000001, 0x0000000000008082, 0x800000000000808a,
+    0x8000000080008000, 0x000000000000808b, 0x0000000080000001,
+    0x8000000080008081, 0x8000000000008009, 0x000000000000008a,
+    0x0000000000000088, 0x0000000080008009, 0x000000008000000a,
+    0x000000008000808b, 0x800000000000008b, 0x8000000000008089,
+    0x8000000000008003, 0x8000000000008002, 0x8000000000000080,
+    0x000000000000800a, 0x800000008000000a, 0x8000000080008081,
+    0x8000000000008080, 0x0000000080000001, 0x8000000080008008
+];
+
+// ln 35 - 40
+// change endianness for a 32 bit number byte by byte
+// e.g. 0xaabbccdd -> 0xddccbbaa
+let swap_u32: int -> int = |x| 
+    or((or(and((x << 8), 0xff00ff00), and((x >> 8), 0x00ff00ff)) >> 16),
+    (or(and((x << 8), 0xff00ff00), and((x >> 8), 0x00ff00ff)) << 16)); 
+
+// compression function
+
+// ln 44
+let bc: int[5] = [0, 0, 0, 0, 0];
+
+// ln 45
+let t: int = 0;
+
+// ln 47 - 49
+let swap_u32_loop: int[25] -> int[25] = |st| array::map(array::new(25, |i| i), |i| swap_u32(st[i]));
+
+// ln 53 - 55
+let theta_bc: int[25], i -> int = |st, i| xor(st[i], xor(st[i + 5], xor(st[i + 10], xor(st[i + 15], st[i + 20]))));
```

Comment:
To improve readability, we could also introduce a helper function
```suggestion
let xor_mult: int[] -> int = |input| std::array::fold(input, 0, |x, y| xor(x, y));
let theta_bc: int[25], i -> int = |st, i| xor_mult([st[i], st[i + 5], st[i + 10], st[i + 15], st[i + 20]]);
```

Reply:
Fixed